### PR TITLE
feat: Snowflake-to-DuckDB data type translation and driver compatibility fixes

### LINKED ISSUE #45: Type Translation
I'm trying to use this emulator in some integration tests and I'm running into issues with type translation that would require me to have to make changes to my application.  The one's that I've run into so far are:

| Snowflake | DuckDB |
|------------|-----------|
| NUMBER | NUMERIC |
| TEXT | VARCHAR |
| TIMESTAMP_NTZ or TIMESTAMP_LTZ | TIMESTAMP |
|  TIMESTAMP_TZ | TIMESTAMP WITH TIME ZONE |

Have you considered adding support for this? I'm a newbie with Go but I can try to put together a PR if you'd be interested in accepting it.

### PR DESCRIPTION
**Pull request by roend83**
## Summary

- Add Snowflake-to-DuckDB data type translation for DDL statements (`CREATE TABLE`, `ALTER TABLE`) and CAST expressions, enabling real Snowflake DDL to work against the emulator
- Fix `INFORMATION_SCHEMA` queries by stripping backticks added by vitess-sqlparser
- Fix DDL/DML response format to include empty `rowtype`/`rowset` arrays for .NET driver compatibility
- Use Snowflake wire protocol type names (`FIXED`, `REAL`) instead of SQL-level names (`NUMBER`, `FLOAT`) in column metadata

### Data type translation (17 mappings)
| Snowflake | DuckDB |
|---|---|
| `NUMBER` | `NUMERIC` |
| `TEXT`, `STRING`, `CHAR`, `CHARACTER` | `VARCHAR` |
| `TIMESTAMP_NTZ` | `TIMESTAMP` |
| `TIMESTAMP_LTZ`, `TIMESTAMP_TZ` | `TIMESTAMPTZ` |
| `VARIANT`, `OBJECT`, `ARRAY` | `JSON` |
| `BINARY`, `VARBINARY` | `BLOB` |
| `FLOAT4` | `FLOAT` |
| `FLOAT8` | `DOUBLE` |
| `BYTEINT` | `TINYINT` |
| `DATETIME` | `TIMESTAMP` |

DDL uses word-boundary-aware string replacement with string literal protection. DML targets only `convert()` expressions to avoid false positives on column names matching type names.

### INFORMATION_SCHEMA fix
vitess-sqlparser backtick-quotes MySQL reserved words (e.g., `` INFORMATION_SCHEMA.`tables` ``). DuckDB rejects backtick quoting, so they are stripped during post-processing.

### Driver compatibility fixes
- Some Snowflake drivers (e.g., the .NET `Snowflake.Data` driver) expect `rowtype` and `rowset` fields in every query response. Removed `omitempty` from JSON tags and initialize as empty arrays for DDL/DML.
- The Snowflake wire protocol uses `FIXED` for numeric types and `REAL` for floating-point, not `NUMBER`/`FLOAT`. Updated the type mapper to match.

## Test plan
- [x] 43 new unit tests covering DDL, CAST, ALTER, INFORMATION_SCHEMA, and edge cases
- [x] All existing tests pass (`go test ./...`)
- [x] End-to-end testing with all three examples (gosnowflake, restapi, docker)

Closes #45

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by roend83** [comment]
@nnnkkk7 I ended up putting a few different things in this PR based on the needs of my current integration testing suite:

1. Support for Snowflake to DuckDB type differences
2. Support for querying the INFORMATION_SCHEMA tables
3. Fixed some API issues that caused the .NET snowflake driver to complain
4. Added support for the snowflake CREATE DATABASE query instead of calling the create database API

Claude generated all of this code for me.  It looked reasonable to me but I'm also not very proficient in Go so I'm happy to make any changes you think are necessary.

**Comment by roend83** [comment]
I reverted the change to support CREATE DATABASE commands after I realized that this emulator doesn't really support multiple databases (it's translating database creations to creating new schemas).

I also added several more changes to get the emulator working with the .NET snowflake driver.  I think these are all changes to better emulate realistic snowflake responses.  I don't believe I need any other changes to get my integration test suites working.

**Review by nnnkkk7 — COMMENTED** [review]
@roend83 Sorry for the slow turnaround on this review, and thank you so much for the work here.

File: pkg/query/translator.go [hunk 1]
```diff
@@ -258,6 +281,254 @@ func removeDualSuffix(sql string) string {
...
+		{"ARRAY", "JSON"},
+		// 4 chars
+		{"TEXT", "VARCHAR"},
+		{"CHAR", "VARCHAR"},
+	}
+}
+
+// translateDataTypes replaces Snowflake type names with DuckDB equivalents in SQL text.
+// It protects string literals from replacement and uses word-boundary-aware matching.
+func (t *Translator) translateDataTypes(sql string) string {
```
Comment (nnnkkk7) [inline pkg/query/translator.go:319]: This rewrites any word matching a type keyword, including identifiers — `DROP TABLE text` → `DROP TABLE VARCHAR`, `CREATE TABLE t (number INTEGER)` → `NUMERIC INTEGER`. Common column/table names will silently break.

File: pkg/query/translator.go [hunk 2]
```diff
@@ -258,6 +281,254 @@ func removeDualSuffix(sql string) string {
...
+	}
+
+	// Restore string literals
+	return restoreStringLiterals(protected, literals)
+}
+
+// translateCastTypes translates Snowflake type names only within convert() expressions.
+// vitess-sqlparser converts CAST(x AS TYPE) to convert(x, TYPE), so we target that pattern.
+// This avoids false positives where column names happen to match type names (e.g., "text").
+func (t *Translator) translateCastTypes(sql string) string {
```
Comment (nnnkkk7) [inline pkg/query/translator.go:335]: No string-literal protection here (unlike `translateDataTypes`), so `CAST('unmatched (' AS NUMBER)` throws off the paren-depth scan and leaves `NUMBER` untranslated. Consider reusing `protectStringLiterals` here too.

File: pkg/query/translator.go [hunk 3]
```diff
@@ -258,6 +281,254 @@ func removeDualSuffix(sql string) string {
...
+					depth--
+				}
+				end++
+			}
+
+			if depth == 0 {
+				// args = everything between convert( and )
+				args := sql[start : end-1]
+				// Find the last comma (separating expr from type)
+				lastComma := strings.LastIndex(args, ",")
```
Comment (nnnkkk7) [inline pkg/query/translator.go:363]: Splits on the *last* comma to find the type, which breaks parameterized types: `CAST(x AS NUMBER(10,2))` → `convert(x, NUMBER(10, 2))` (untranslated, invalid DuckDB type). Should split on the first top-level comma instead.

File: pkg/query/translator.go [hunk 4]
```diff
@@ -258,6 +281,254 @@ func removeDualSuffix(sql string) string {
...
+				if lastComma >= 0 {
+					expr := args[:lastComma]
+					typePart := strings.TrimSpace(args[lastComma+1:])
+					// Translate the type name
+					translatedType := t.translateSingleType(typePart)
+					result.WriteString(sql[i : i+8])
+					result.WriteString(expr)
```
Comment (nnnkkk7) [inline pkg/query/translator.go:370]: Only the outer `convert()` gets translated — nested CASTs leave their inner type untranslated: `CAST(CAST(a AS NUMBER) AS TEXT)` → `convert(convert(a, NUMBER), VARCHAR)`. Needs recursive handling of `expr`.

File: pkg/query/translator.go [hunk 5]
```diff
@@ -258,6 +281,254 @@ func removeDualSuffix(sql string) string {
...
+	for _, m := range t.typeMappings {
+		if upper == m.from {
+			return m.to
+		}
+	}
+	return typeName
+}
+
+// protectStringLiterals replaces single-quoted string literals with placeholders
+// to prevent type name replacement inside strings.
```
Comment (nnnkkk7) [inline pkg/query/translator.go:398]: `protectStringLiterals` and `stripBacktickIdentifiers` duplicate the same string-literal-scanning logic. Worth extracting a shared helper to avoid future drift (low priority).
